### ISSUE TITLE
strange behavior List#corresponds

### ISSUE THREAD

**Issue by xuwei-k** [scala-native/scala-native#695]
version 0.2

```scala
trait Equal[A] {
  def equal(x: A, y: A): Boolean
}

object Equal {
  implicit def int: Equal[Int] =
    new Equal[Int] {
      def equal(x: Int, y: Int): Boolean = x == y
    }

  implicit def string: Equal[String] =
    new Equal[String] {
      def equal(x: String, y: String): Boolean = x == y
    }

  implicit def list[A](implicit A: Equal[A]): Equal[List[A]] =
    new Equal[List[A]] {
      def equal(x: List[A], y: List[A]): Boolean =
        (x corresponds y)(A.equal(_, _))
    }
}

object Main {
  def main(args: Array[String]): Unit = {
    println(implicitly[Equal[Int]].equal(42, 42)) // true

    println(implicitly[Equal[String]].equal("a", "a")) // true

    val a = 42 :: Nil
    // false in scala-native!? true in JVM.
    println(implicitly[Equal[List[Int]]].equal(a, a))

    val b = "xyz" :: Nil
    println(implicitly[Equal[List[String]]].equal(b, b)) // false !?
  }
}
```

**Comment by densh** [comment]
Definitely a bug, thanks for reporting @xuwei-k !

**Comment by xuwei-k** [comment]
more simple example

```scala
object Main {
  def main(args: Array[String]): Unit = {
    val a = List(1, 2, 3)

    // false in scala-native, true in JVM
    println(a.corresponds(a)(_ == _))
  }
}
```

**Comment by xuwei-k** [comment]
```scala
object Main {
  def main(args: Array[String]): Unit = {
    val a = List(1, 2, 3)

    val result = a.corresponds(a){ (x, y) =>
      println((x, y))
      true
    }

    println(result)
  }
}
```


```
(1,1)
(1,2)
(1,3)
false
```

:scream:

**Comment by densh** [comment]
@xuwei-k Thanks for the minimized example!

**Comment by densh** [comment]
I was able to find the root cause for this one: front-end incorrectly resolves symbols to SSA names in the front-end for methods that were transformed by tailrec phase. Instead of using the internal loop's version of `this` variable, it always uses the original `this`, never advancing to the tail. Should be easy to fix in 0.3.

**Comment by sjrd** [comment]
See https://github.com/scala-js/scala-js/blob/f21e7f5ea652fe4a186f1ecd8c0070fbb80edc80/compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala#L1966-L1981, and more generally references to `thisLocalVarIdent` in that file.
